### PR TITLE
Feature reduced hierarchy transform update

### PR DESCRIPTION
**Pull request by rams3s**
This change is [Reviewable](https://reviewable.io/reviews/fishingcactus/openfl/118)

### CODE REVIEW

File: format/swf/lite/MovieClip.hx [hunk 1]
```diff
@@ -710,14 +710,13 @@ class MovieClip extends flash.display.MovieClip {
 
 	}
 
-	public override function __update (transformOnly:Bool, updateChildren:Bool, ?maskGraphics:Graphics = null):Void {
-
-		super.__update(transformOnly, updateChildren, maskGraphics);
+	private function __update9SliceBitmap ():Void {
 
 		// :TODO: should be in a prerender phase
 		// :TODO: use dirty flag if need to update __9SliceBitmap
```
Comment (NuclearCookie) [inline format/swf/lite/MovieClip.hx:717]: you adressed these todos

File: openfl/display/BitmapData.hx [hunk 2]
```diff
@@ -412,7 +422,7 @@ class BitmapData implements IBitmapDrawable {
 
 		if (colorTransform != null) {
 			
-			var copy = new BitmapData (Reflect.getProperty (source, "width"), Reflect.getProperty (source, "height"), true, 0);
+			var copy = new BitmapData (Reflect.getProperty (source, "__width"), Reflect.getProperty (source, "__height"), true, 0);
```
Comment (NuclearCookie) [inline openfl/display/BitmapData.hx]: use Reflect.getField now instead of getProperty

File: openfl/display/BitmapData.hx [hunk 3]
```diff
@@ -496,7 +506,7 @@ class BitmapData implements IBitmapDrawable {
 		
 		if (colorTransform != null) {
 			
-			var copy = new BitmapData (Reflect.getProperty (source, "width"), Reflect.getProperty (source, "height"), true, 0);
+			var copy = new BitmapData (Reflect.getProperty (source, "__width"), Reflect.getProperty (source, "__height"), true, 0);
```
Comment (NuclearCookie) [inline openfl/display/BitmapData.hx]: same here

**Review by NuclearCookie — CHANGES_REQUESTED** [review]

**Comment by NuclearCookie** [comment]
Also, When reusing a 9slice graphic, I am experiencing issues with transparency. ( state does not seem to be reset ).

**Comment by NuclearCookie** [comment]
Blend modes also seem broken ( multiply )

File: format/swf/lite/MovieClip.hx [hunk 4]
```diff
@@ -710,14 +710,13 @@ class MovieClip extends flash.display.MovieClip {
...
 
 		if (__symbol != null && __symbol.scalingGridRect != null && __9SliceBitmap == null) {
```
Comment (NuclearCookie) [inline format/swf/lite/MovieClip.hx:719]: This condition can be avoided if you look at the place this function is called

File: openfl/_internal/renderer/RenderSession.hx [hunk 5]
```diff
@@ -42,13 +43,57 @@ class RenderSession {
...
 	
 	public function new () {
 		
 		//maskManager = new MaskManager (this);
-		
+		renderTargetBaseTransformStack = new GenericStack<Matrix> ();
+		pushRenderTargetBaseTransform (@:privateAccess Matrix.__identity, null);
 	}
 	
+	public function pushRenderTargetBaseTransform (source:Dynamic, renderTargetBaseTransform:Matrix) {
```
Comment (NuclearCookie) [inline openfl/_internal/renderer/RenderSession.hx]: Source could be IBitmapDrawable

File: format/swf/lite/MovieClip.hx [hunk 6]
```diff
@@ -736,16 +736,14 @@ class MovieClip extends flash.display.MovieClip {
...
-				__renderTransform.translate(-bounds.x, -bounds.y);
-				Rectangle.pool.put(bounds);
-				@:privateAccess bitmap.__drawGL(renderSession, this, __renderTransform);
-
-				__renderTransform = previousTransform;
+				var renderTransform = Matrix.pool.get ();
+				renderTransform.identity ();
+				renderTransform.translate(-bounds.x, -bounds.y);
+				@:privateAccess bitmap.__drawGL(renderSession, this, renderTransform);
+				Matrix.pool.put (renderTransform);
```
Comment (NuclearCookie) [inline format/swf/lite/MovieClip.hx:746]: re-add the bounds to the pool

**Comment by NuclearCookie** [comment]
I am experiencing issues with cached as bitmap. If I use it, masks seem to be ignored. 
First thought:
When drawing the mask, we used to update all the children https://github.com/FishingCactus/openfl/pull/118/files#diff-2b58ec07f66353241b751100e39589a9L181
This will have to be added somewhere

File: format/swf/lite/MovieClip.hx (hunk 4 above)
Comment (rams3s) [reply format/swf/lite/MovieClip.hx:719]: not exactly the same test

File: format/swf/lite/MovieClip.hx (hunk 6 above)
Comment (rams3s) [reply format/swf/lite/MovieClip.hx:746]: good catch ;)

File: openfl/display/BitmapData.hx (hunk 2 above)
Comment (rams3s) [reply openfl/display/BitmapData.hx]: true

File: openfl/display/BitmapData.hx (hunk 3 above)
Comment (rams3s) [reply openfl/display/BitmapData.hx]: true again

File: openfl/_internal/renderer/RenderSession.hx (hunk 5 above)
Comment (rams3s) [reply openfl/_internal/renderer/RenderSession.hx]: done

**Comment by NuclearCookie** [comment]
Review status: 0 of 24 files reviewed at latest revision, 5 unresolved discussions.

---

*[format/swf/lite/MovieClip.hx, line 139 at r11](https://reviewable.io:443/reviews/fishingcactus/openfl/118#-K_j0EnXsTSxRH646Amy:-K_j0EnXsTSxRH646Amz:bmetkql) ([raw file](https://github.com/fishingcactus/openfl/blob/155f8c632069518db6cfee8dfcc22985e132ea7c/format/swf/lite/MovieClip.hx#L139)):*
> if (!__useSeparateRenderScaleTransform) {
> 			
> 			for( i in 0 ... numChildren ){
> 				var child = getChildAt(i);
> 				child.__useSeparateRenderScaleTransform = false;
> 			}
> 			
> 		}
>  

What about adding a child using AddChild ? Don't we have to do the same check at that moment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/118)*

**Comment by rams3s** [comment]
Review status: 0 of 27 files reviewed at latest revision, 5 unresolved discussions.

---

*[format/swf/lite/MovieClip.hx, line 716 at r1](https://reviewable.io:443/reviews/fishingcactus/openfl/118#-KZ7YFfRCclnsHq5FDmu-r1-716:-KaC6y2w-9YV91rkWAi5:b-896fix) ([raw file](https://github.com/fishingcactus/openfl/blob/ee1fed5116e51c3d60a686020003cbc516600955/format/swf/lite/MovieClip.hx#L716)):*
<details><summary><i>Previously, NuclearCookie (Pieter Vantorre) wrote…</i></summary><blockquote>

you adressed these todos
</blockquote></details>

Done.

---

*[format/swf/lite/MovieClip.hx, line 139 at r11](https://reviewable.io:443/reviews/fishingcactus/openfl/118#-K_j0EnXsTSxRH646Amy:-KaBqTf6KRWyC3iraVqu:b-hm308m) ([raw file](https://github.com/fishingcactus/openfl/blob/155f8c632069518db6cfee8dfcc22985e132ea7c/format/swf/lite/MovieClip.hx#L139)):*
<details><summary><i>Previously, NuclearCookie (Pieter Vantorre) wrote…</i></summary><blockquote>

> if (!__useSeparateRenderScaleTransform) {
> 			
> 			for( i in 0 ... numChildren ){
> 				var child = getChildAt(i);
> 				child.__useSeparateRenderScaleTransform = false;
> 			}
> 			
> 		}
>  

What about adding a child using AddChild ? Don't we have to do the same check at that moment?

</blockquote></details>

Shouldn't happen (__useSeparateRenderScaleTransform is false for Bitmaps and 9 slice shapes only) but moved this code to addChild for safety. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/118)*

**Pull request merged**